src/shogun/modelselection/parameter_tree.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_PARAMETER_TREE_H
+#define SHOGUN_PARAMETER_TREE_H
+
+#include <shogun/base/AnyParameter.h>
+#include <shogun/base/SGObject.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/type_case.h>
+#include <shogun/machine/Machine.h>
+#include <unordered_set>
+
+namespace shogun
+{
+
+	class ParameterNode : public CSGObject
+	{
+
+	public:
+		/**
+		 * Parameter node constructor with model
+		 * @param model
+		 * @param param_properties
+		 */
+		ParameterNode(CSGObject& model);
+
+		ParameterNode(CSGObject& model, ParameterProperties param_propeties);
+
+#ifndef SWIG
+		/**
+		 * The default constructor
+		 */
+		ParameterNode() = default;
+
+		/**
+		 * Creates a node that mimics the CSGObject structure
+		 * and places it in the tree using the given name.
+		 * @param name
+		 * @param obj
+		 */
+		virtual void create_node(const std::string& name, CSGObject* obj);
+#endif
+
+		/**
+		 * Attach a new node at a given location in the tree
+		 * defined by the parameter name
+		 * @param name
+		 * @param node
+		 */
+		ParameterNode* attach(
```

Comment:
I would hide most things from swig in here, e.g. this one

Reply:
This will be needed to add another GridSearch instance